### PR TITLE
Add subscription support

### PR DESCRIPTION
**Pull request by madeye**
- [x] Add animations
- [x] Fix Icons
- [x] Handle exceptions nicely
- [x] Check race conditions

### CODE REVIEW

**Comment by Mygod** [comment]
This seems overly complicated. Why do we need multiple subscriptions anyway?

**Comment by madeye** [comment]
My understanding is that we need a separate fragment to manage the subscriptions.

Not sure if users would use multiple subscriptions. But I didn't see multiple subscriptions would bring much complexity.

**Comment by Mygod** [comment]
I see. Sure then.

**Comment by madeye** [comment]
@Mygod I'm not sure if there would be a race condition that users try to modify profiles when downloading profiles from remote.

Or maybe we can prevent users switch to other fragment when downloading the profiles.

**Comment by Mygod** [comment]
There will not be any races if you simply do everything in main. :)

**Comment by madeye** [comment]
I'm using `GlobalScope.launch` to fetch profiles and apply them, that's what I'm worry about...

**Comment by madeye** [comment]
Any thoughts?

**Comment by Mygod** [comment]
Sorry I'll look at it hopefully later today.

**Comment by madeye** [comment]
I plan to merge this change and schedule a pre-release soon.

Please comment if there's any concern.

**Comment by Mygod** [comment]
My concern is that these subscriptions does not quite place nicely with profiles that are added manually. Some possible improvements:

* Flag profiles that are added by subscriptions and make their server settings immutable (but copyable and deletable);
* Respect existing (subscription) profile's feature settings when replacing; (priorities for choosing feature settings: existing not-necessarily-subscription profile -> JSON profile -> currently selected profile)
* When syncing, hide (as opposed to removing) profiles that are no longer present. (in other words, the only way to re-sync feature settings is to remove them manually and re-sync)

@madeye Thoughts?

**Comment by madeye** [comment]
I agree that we can add a new flag to each profile, to mark it whether managed by subscription. If a profile is "managed", then it should be immutable and can be overwritten by each sync.

**Comment by madeye** [comment]
@Mygod Can you help add this "managed" profile feature?

I am not very familiar with the current database design, but the improvement here requires some change to Room related code.

**Comment by Mygod** [comment]
Sure but I probably will not have time to test it until weekend. Room is very similar to ORMlite.

File: mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt [hunk 1]
```diff
@@ -590,8 +591,12 @@ class ProfilesFragment : ToolbarFragment(), Toolbar.OnMenuItemClickListener {
         profilesAdapter.deepRefreshId(profileId)
     }
 
-    override fun onDestroyView() {
+    override fun onPause() {
         undoManager.flush()
+        super.onPause()
+    }
+
+    override fun onDestroyView() {
```
[the remark this answers is not in the document]
Comment (madeye) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: Without flushing undoManager, we would have a crash after switching fragments.

Removing a profile -> switch to subscription fragment -> sync

**Comment by madeye** [comment]
I can help do the testing. Or you can just add the new field in profile, then I can implement all the other logic.

**Comment by Mygod** [comment]
@madeye Done.

File: mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mygod) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: I thought switching fragments will trigger `onDestroyView`, no?

[the remark this answers is not in the document]
Comment (madeye) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: According to my local test, the answer is no...

[the remark this answers is not in the document]
Comment (madeye) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: Okay, it indeed trigger `onDestroyView`.

The actually problem is the `snackbar.dismiss()` is async and may happen after subscription updating.

Is there anyway to ensure undoManager is committed before switching fragment?

[the remark this answers is not in the document]
Comment (Mygod) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: I'm not sure. Why exactly would that be a problem?

[the remark this answers is not in the document]
Comment (Mygod) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: This patch should dismiss synchronously -- though I am not exactly sure why that is needed.

```kotlin
--- a/mobile/src/main/java/com/github/shadowsocks/widget/UndoSnackbarManager.kt
+++ b/mobile/src/main/java/com/github/shadowsocks/widget/UndoSnackbarManager.kt
@@ -32,21 +32,23 @@ import com.google.android.material.snackbar.Snackbar
  * @tparam T Item type.
  */
 class UndoSnackbarManager<in T>(private val activity: MainActivity, private val undo: (List<Pair<Int, T>>) -> Unit,
-                                commit: ((List<Pair<Int, T>>) -> Unit)? = null) {
+                                private val commit: ((List<Pair<Int, T>>) -> Unit)? = null) {
     private val recycleBin = ArrayList<Pair<Int, T>>()
     private val removedCallback = object : Snackbar.Callback() {
         override fun onDismissed(transientBottomBar: Snackbar?, event: Int) {
             when (event) {
-                DISMISS_EVENT_SWIPE, DISMISS_EVENT_MANUAL, DISMISS_EVENT_TIMEOUT -> {
-                    commit?.invoke(recycleBin)
-                    recycleBin.clear()
-                }
+                DISMISS_EVENT_SWIPE, DISMISS_EVENT_TIMEOUT -> doCommit()
             }
             last = null
         }
     }
     private var last: Snackbar? = null
 
+    private fun doCommit() {
+        commit?.invoke(recycleBin)
+        recycleBin.clear()
+    }
+
     fun remove(items: Collection<Pair<Int, T>>) {
         recycleBin.addAll(items)
         val count = recycleBin.size
@@ -62,6 +64,7 @@ class UndoSnackbarManager<in T>(private val activity: MainActivity, private val
     fun remove(vararg items: Pair<Int, T>) = remove(items.toList())
 
     fun flush() {
+        doCommit()
         last?.dismiss()
         last = null
     }
```

[the remark this answers is not in the document]
Comment (madeye) [reply mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt]: Cool! The problem solved.

**Comment by madeye** [comment]
Current design:

1. Any profile synced from subscription is marked as "managed".
2. All the server info fields in a managed profile is immutable.
3. Managed profiles will be deleted from local if it's not present in the latest remote config.

**Comment by madeye** [comment]
Merged to the master branch now.

**Comment by Mygod** [comment]
Okay. I'll check it out later today (am travelling).

**Pull request merged**